_posts/2023-06-27-redis.md

```diff
@@ -0,0 +1,23 @@
+ 
+These facts only hold when the size of your payload and the number of connections remain relatively small. This easily jumps out the window with ever-increasing load parameters. The threshold is, unfortunately, rather low at a high number of connections and increased payload sizes. Modern large-scale micro-services will easily have over 100 running instances at medium scale. And since most instances employ some pooling mechanism so as not to pay a connection cost for each request, a single Redis instance is going to do a bit of work in maintaining those connections, not to talk of serving requests as they come through. To improve performance at medium to high loads, some projects such as [KeyDb](https://docs.keydb.dev), Snapchat's drop-in replacement as an alternative to Redis, employ a multithreaded approach and a bit of magic to sustain some high workloads. This has been touted to provide 5x performance over Redis. Another solution which I have seen used is to employ a proxy that multiplexes over multiple redises. One such proxy is the twoemproxy developed at Twitter. [Twoemproxy](https://github.com/twitter/twemproxy), or nutcracker as it is informally known, is in itself single-threaded and employs key hashing to store keys in shards of multiple redis instances, giving you proper multiplexing. While it may look like this is susceptible to the original problems of a single-threaded application, it is not necessarily the reality because twoemproxy employs a single thread for each Redis instance, turning the whole system into a multi-threaded system. Of course, this is still susceptible to hotkeys. These might look like ideal solutions, but setting up new infrastructure as an intermediary service introduces a new failure point, which is neither trivial nor ideal. But when done right, there are a lot of net positives. 
```

Comment:
Good overall but you might want to talk about [redis-cluster](https://redis.io/docs/management/scaling/) or [master slave](https://redis.io/docs/management/replication/) coupled with [redis-sentinel](https://redis.io/docs/management/sentinel/) as  a way to horizontally scale redis deployments and its downsides.

Reply:
i delibrately avoided them because people hardly use them in the realworld and durability is not guaranteed even in cluster mode